### ISSUE TITLE
REST: get things of type

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#4302]
There is currently no method to get all things of a special type only (or I didn't find them).
We are using setups that "get all things" call results into a > 2 MiB response.
Sometimes we need to know only things of a given type first.
I can create that interface in our custom product or you could extend the thing resource to support this request officially.
WDYT?

**Comment by sjka** [comment]
I don't see a reason _not_ to put it here. "go" from my side.

**Comment by kaikreuzer** [comment]
We didn't go into detailed filtering mechanisms on any rest resource from the start, but the it was clear that this would be required sooner or later.

When adding those now for certain cases, we should make sure to find a good general scheme that we can apply to all different kinds of resources. Ideally, we should have flexible query parameters, maybe also support substring matching and somehow cleverly feed that into a stream-based get method on the registries.

**Comment by maggu2810** [comment]
The `GET` method does not support data in the content, so we could use the URL for the filter information. Correct?

**Comment by htreu** [comment]
Correct. We could provide something like `/rest/things?thing-type=hue:lamp` which will make `thing-type` a pre defined query parameter for the things resource. The alternative is to encode the whole filter into a single parameter like `rest/things?filter=thing-type%3Dhue%3Alamp%3Blabel%3DMy%20Hue%20Lamp` which would be more flexible in a way that we create filters dynamically from the filter string. This way nearly every attribute or property of a resource may be used as a filter.

**Comment by maggu2810** [comment]
We would need a implementation that could be reused e.g.

```Java
boolean testLdapExpression(final String expression, final Map<String, String> map)
```

The `map` is filled with key value pairs and is under control of the caller.
The thing resource could handle it this way (pseudo code):

```Java
for (final ThingDTO thing : getAllThings()) {
    final Map<String,String> props = new HashMap<>();
    props.put("thing-type", thing.thingType);
    props.put("label", thing.label);
    if (testLdapExpression(filter, props) {
        result.add(thing);
    }
}
```

another implementation will put other properties...

**Comment by kaikreuzer** [comment]
I am personally not a huge fan of LDAP filters and imho this is also rather unusual in the JS/REST world - but maybe we should ask some web experts on that :-)

**Comment by htreu** [comment]
What I like better on the explicit query parameter name (i.e. `/rest/things?thing-type=hue:lamp`) is the fact that the swagger documentation will give you direct access to all available parameters. In addition the URI is more human readable and I suspect we will not have that many query params.

**Comment by maggu2810** [comment]
I "like" the query parameter, too as it is more readable.

> we should make sure to find a good general scheme that we can apply to all different kinds of resources. Ideally, we should have flexible query parameters, maybe also support substring matching 

The query parameter seems to be not very flexible. At least it is not clear to me how to differ between a "logical and" and a "logical or". Perhaps we can agree that the parameters are always read as logical and / or, or it could be at least a global combination defined.

**Comment by sjka** [comment]
In that respect the query parameters look like a dead-end. I mean, the don't need to be extremely powerful, but having both "and" _and_ "or" sounds feasible. 

I personally also don't really LDAP expressions, as they are hard to read & write for humans, despite the fact that they are incredibly simple to parse for computers which would be a big plus on the positive side. 

However, it also somehow feels wrong to invent our own. If it's not gonna be LDAP, how about using some other popular one, e.g. the filter language from [OData](https://msdn.microsoft.com/library/gg309461.aspx#BKMK_filter)? It's basically what anybody would write in code in almost every language, except that the special characters are replaced by abbreviations which are already familiar from e.g. HTML.

**Comment by htreu** [comment]
I agree on the fact that specific query parameters may become too basic very quick. Going with the OData filter language looks promissing. Although we have to separate the OData protocol with lots of other resource defining specs from just the query/filter part.
[Apache Olingo](https://olingo.apache.org/) does provide an implementation but from the first look it seems to be bloated when we just want the filter stuff to work. And libs are not broken into specific aspects of OData but client/server only...
Nevertheless its a good proposal and we should further investigate implementation possibilities.
Maybe this is a "help wanted" topic? wdyt?

**Comment by maggu2810** [comment]
> If it's not gonna be LDAP, how about using some other popular one

Why not using the RFC 1960-based filter that is used for ServiceReference matches already by the OSGi framework. I assume that filters are (at least for OSGi bundle developers) popular (but hey isn't that LDAP :wink:).

* [BundleContext#createFilter]
* [Filter]

On the OSGi container side it would come without any overhead (I assume that nearly all products are using components).
And "too complicated" shouldn't be a problem. The UI don't need to support the building of complicated filters.

```Java
final Filter filter = bundleContext.createFilter(userFilter);
final Stream<EnrichedThingDTO> thingStream = thingRegistry.stream().filter(thing -> {
    final Dictionary<String, Object> dict = new Hashtable<>();
    dict.put("thingTypeUID", thing.getThingTypeUID().getAsString());
    dict.put("label", thing.getLabel());
    dict.put("location", thing.getLocation());
    return filter.match(dict);
}).map(t -> convertToEnrichedThingDTO(t, locale)).distinct();
```

[BundleContext#createFilter]: https://osgi.org/javadoc/r4v43/core/org/osgi/framework/BundleContext.html#createFilter(java.lang.String)
[Filter]: https://osgi.org/javadoc/r4v43/core/org/osgi/framework/Filter.html

**Comment by maggu2810** [comment]
I have added a simple demonstration for my suggestion, see [diff]

You can use e.g.
* `http://127.0.0.1:8080/rest/things/filtered/(thingTypeUID=magic:*)` to get all things using a thing type UID of the magic bundle
* `http://127.0.0.1:8080/rest/things/filtered/(!(thingTypeUID=magic:*))` to get all things using a thing type UID not from the magic bundle

I used that exmaples to demonstrate that wildcards could be used, too.

IMHO which complex filters (AND, OR, NOT, wildcards, ...) are offered by the UI should be UI specific. At least the REST API itself, that could be used by "whatever" should be generic and this is a very simple solution as in an OSGi runtime all that magic filter handling is already present.

[diff]: https://github.com/eclipse/smarthome/compare/master...maggu2810:issue4302?expand=1

**Comment by RobWin** [comment]
Hello,

please keep in mind that the LDAP-filter notation contains URL-unsafe characters which must be escaped by clients.
There is another interesting notation called [RSQL ](https://github.com/jirutka/rsql-parser)(Rest Resource Query Language) which is based on [FIQL](https://tools.ietf.org/html/draft-nottingham-atompub-fiql-00) (Feed Item Query Language) – an URI-friendly syntax for expressing filters across the entries in an Atom Feed. RSQL provides a friendlier syntax for logical operators and some of the comparison operators.

Apache CXF provides a search functionality based on FIQL. See http://cxf.apache.org/docs/jax-rs-search.html#JAX-RSSearch-FIQL
